File: js/SinCos_test.js
import {assertEquals} from 'goog:goog.testing.asserts';
import 'goog:goog.testing.jsunit';
import * as Crc64 from './Crc64.js';
import * as SinCos from './SinCos.js';

function testSin() {
  let /** @type{!Uint32Array} */ crc = Crc64.init();
  for (let /** @type{number} */ i = 0; i < SinCos.MAX_ANGLE; ++i) {
    console.log(SinCos.SIN[i]);
    crc = Crc64.update(crc, SinCos.SIN[i] & 0xFF);
  }
  assertEquals('9486473C3841E28F', Crc64.finish(crc));
}

function testCos() {
  let /** @type{!Uint32Array} */ crc = Crc64.init();
  for (let /** @type{number} */ i = 0; i < SinCos.MAX_ANGLE; ++i) {
    crc = Crc64.update(crc, SinCos.COS[i] & 0xFF);
  }
  assertEquals('A32700985A177AE9', Crc64.finish(crc));
}

File: c/sin_cos_test.cc
#include "sin_cos.h"

#include <cmath>

#include "crc64.h"
#include "gtest/gtest.h"

namespace twim {

TEST(SinCosTest, Sin) {
  uint64_t crc = Crc64::init();
  for (size_t i = 0; i < SinCos.kMaxAngle; ++i) {
    crc = Crc64::update(crc, SinCos.kSin[i] & 0xFF);
  }
  EXPECT_EQ("9486473C3841E28F", Crc64::finish(crc));
}

TEST(SinCosTest, Cos) {
  uint64_t crc = Crc64::init();
  for (size_t i = 0; i < SinCos.kMaxAngle; ++i) {
    crc = Crc64::update(crc, SinCos.kCos[i] & 0xFF);
  }
  EXPECT_EQ("A32700985A177AE9", Crc64::finish(crc));
}

TEST(SinCosTest, Log2) {
  for (size_t i = 1; i < 64; ++i) {
    float diff = (std::log(i) / std::log(2)) - SinCos.kLog2[i];
    EXPECT_LE(std::abs(diff), 0.00000022f) << i;
  }
}

// TODO(eustas): add Pow2 test

}  // namespace twim
